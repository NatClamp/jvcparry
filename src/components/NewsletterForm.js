import React, { useState, useEffect } from 'react';
import { Input, Button, Text } from 'atomize';
import { useForm } from 'react-hook-form';
import jsonp from 'jsonp';
import toQueryString from 'to-querystring';
import PropTypes from 'prop-types';

const NewsletterForm = (props) => {
  const {
    register, handleSubmit, reset, errors, formState: { isSubmitSuccessful },
  } = useForm();
  const [mailchimpMessage, setMailchimpMessage] = useState(null);
  const [error, setError] = useState(null);
  const [errorMessage, setErrorMessage] = useState(null);
  const [submittedData, setSubmittedData] = useState({});
  const [successSub, setSuccessSub] = useState(null);

  const { location } = props;

  const onSubmit = (resp) => {
    const emailObj = {
      EMAIL: resp.email_address,
    };
    const params = toQueryString(emailObj);

    jsonp(`${process.env.REACT_APP_MAILCHIMP_URL}&${params}`, { param: 'c' }, (err, data) => {
      if (err) {
        setErrorMessage(err.message);
        setError(true);
        console.log(errorMessage);
      } else {
        console.log(mailchimpMessage);
        setMailchimpMessage(data.msg);
        setSubmittedData(data);
      }
    });
  };

  useEffect(() => {
    if (isSubmitSuccessful) {
      reset({ ...submittedData });
      setSuccessSub(true);
    }
  }, [isSubmitSuccessful, submittedData, reset]);

  return (
    <>
      {location === 'footer' && (
      <Text tag="p" textSize="body" textAlign={{ xs: 'center', md: 'left' }} align={{ xs: 'center' }} justify={{ xs: 'center' }} p={{ y: '5px', l: { xs: '20px', md: '0' }, r: { xs: '20px', md: '0' } }} m={{ xs: '15px auto', md: 'auto' }} textColor="white" data-testid="footer-title">
        Subscribe to my newsletter:
      </Text>
      )}
      <form onSubmit={handleSubmit(onSubmit)}>
        <Input
          name="email_address"
          ref={register({ required: true })}
          type="email"
          placeholder="Email"
          m={{ b: '10px' }}
          suffix={(
            <Button
              pos="absolute"
              hoverBg="hsla(217, 14%, 43%)"
              bg="hsla(217, 14%, 50%)"
              top="0"
              right="0"
              rounded={{ r: 'md' }}
              type="submit"
            >
              Subscribe
            </Button>
          )}
        />
        {successSub && <Text tag="p" textSize="body" textColor={location === 'footer' ? 'white' : 'black'} p={{ t: '10px' }} textAlign={location === 'footer' ? '' : 'center'}>Subscribed!</Text>}
        {errors.emailAddress && <Text tag="p" textSize="body" textColor={location === 'footer' ? 'white' : 'black'} p={{ t: '10px' }} textAlign={location === 'footer' ? '' : 'center'}>This field is required</Text>}
        {error && <Text tag="p" textSize="body" textColor={location === 'footer' ? 'white' : 'black'} p={{ t: '10px' }} textAlign={location === 'footer' ? '' : 'center'}>Sorry, an error occurred. Please try again later.</Text>}
      </form>
    </>
  );
};

NewsletterForm.propTypes = {
  location: PropTypes.string.isRequired,
};

export default NewsletterForm;
