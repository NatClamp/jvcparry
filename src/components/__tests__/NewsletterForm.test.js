/**
 * @jest-environment jsdom
 */

import React from 'react';
import { render, screen, cleanup } from '@testing-library/react';
import { createMemoryHistory } from 'history';
import { Router } from 'react-router-dom';
import renderer from 'react-test-renderer';

import '@testing-library/jest-dom/extend-expect';

import NewsletterForm from '../NewsletterForm';

describe('NewsletterForm component', () => {
  afterEach(cleanup);
  const history = createMemoryHistory();
  it('should render the NewsletterForm component in the footer', () => {
    render(
      <Router history={history}>
        <NewsletterForm location="footer" />
      </Router>,
    );
    expect(screen.getByText(/Subscribe to my newsletter/i)).toBeInTheDocument();
  });
  it('should render the NewsletterForm component in the popup without the title', () => {
    render(
      <Router history={history}>
        <NewsletterForm location="modal" />
      </Router>,
    );
    expect(screen.queryByTestId('footer-title')).not.toBeInTheDocument();
  });
  it('should match the snapshot', () => {
    const tree = renderer
      .create(
        <Router history={history}>
          <NewsletterForm location="footer" />
        </Router>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
