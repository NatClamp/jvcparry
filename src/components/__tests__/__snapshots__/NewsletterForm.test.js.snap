// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewsletterForm component should match the snapshot 1`] = `
Array [
  <p
    className=""
    data-testid="footer-title"
  >
    Subscribe to my newsletter:
  </p>,
  <form
    onSubmit={[Function]}
  >
    <div
      className=""
    >
      <input
        className=""
        d="flex"
        m={
          Object {
            "b": "10px",
          }
        }
        name="email_address"
        placeholder="Email"
        type="email"
      />
      <button
        className=""
        type="submit"
      >
        Subscribe
      </button>
    </div>
  </form>,
]
`;
